[PROPOSED] Web Machine Learning Working Group Charter

anssiko opened the pull request:
This is a work in progress draft charter for the Web Machine Learning Working Group. The charter builds upon the [initial charter](https://www.w3.org/2021/04/web-machine-learning-charter.html). Feel free to review the draft charter and raise [issues](https://github.com/w3c/machine-learning-charter/issues).

The expected timeline is as follows:
- Q4 '22: Charter development
- Q1 '23: W3C Advisory Committee review
- Q2' 23: Charter approved

See also: [Diff](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fwww.w3.org%2F2021%2F04%2Fweb-machine-learning-charter.html&doc2=https%3A%2F%2Fpr-preview.s3.amazonaws.com%2Fw3c%2Fmachine-learning-charter%2Fpull%2F19.html) to the [initial charter](https://www.w3.org/2021/04/web-machine-learning-charter.html).


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/machine-learning-charter/pull/19.html" title="Last updated on Jan 9, 2023, 9:54 AM UTC (183f792)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/machine-learning-charter/19/a033e3a...183f792.html" title="Last updated on Jan 9, 2023, 9:54 AM UTC (183f792)">Diff</a>

anssiko committed:
Syncronize content with the initial charter [1]

[1] https://www.w3.org/2021/04/web-machine-learning-charter.html

anssiko committed:
Remove Khronos from External Organizations

This reflects WG decision to drop WebGL interoperability requirements

anssiko commented:
I have [opened issues](https://github.com/w3c/machine-learning-charter/issues) to track feedback provided during the [kick off call on 6 October 2022](https://www.w3.org/2022/10/06-webmachinelearning-minutes.html#t01).

I will keep this draft PR updated to reflect the WG's consensus that is forming in the issue discussions.

Following this example, feel free to open more issues for feedback that has not yet been recorded. And please provide your comments in the existing open issues to help us prepare our next charter. Thank you for your contributions!

anssiko committed:
Clarify Out of Scope

"base linear algebra operations are out of scope" refers to the
Basic Linear Algebra Subprograms (BLAS) interface.

Related #20

anssiko committed:
Note WebGPU interoperability expectation in Coordination

Fix #20

anssiko committed:
Clarify WebRTC coordination

Confidence level increased.

Fix #21

anssiko committed:
Add high-level to low-level op decompose path

These informative definitions will facilitate progressive lowering
by implementations.

Fix #22
Related #28

anssiko committed:
Expand Motivation and Background and clarify priorities in Scope

Fix #23

anssiko committed:
Add OpenXLA Project to Coordination

Fix #28

anssiko commented:
I pushed an update that addresses the following charter issues:

- #20 
- #21 
- #22 
- #23 
- #28

See also: [Diff](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fwww.w3.org%2F2021%2F04%2Fweb-machine-learning-charter.html&doc2=https%3A%2F%2Fpr-preview.s3.amazonaws.com%2Fw3c%2Fmachine-learning-charter%2Fpull%2F19.html) to the [initial charter](https://www.w3.org/2021/04/web-machine-learning-charter.html).

dontcallmedom reviewed (commented):
* the Timeline section needs an update
* The expected completion date of WebNN needs an update
* have the prospects of the Model API integration in the charter evolved given the recent prototyping efforts?
* the charter should probably link explicitly to the ethics doc now that it exists

Otherwise, I need to check if the charter template has evolved, but the proposed changes themselves look good to me.

anssiko committed:
Address review comments

- Update Expected completion date of WebNN
- Remove now redundant Timeline section
- Link to Ethical Principles for Web Machine Learning WG Note
- Add Adopted Draft and Exclusion Draft placeholders
- Update WebNN and Model Loader Draft state

anssiko commented:
@dontcallmedom thanks! I've pushed an update to address your review comments.

I also incorporated some updates from the evolved charter template, namely Adopted Draft and Exclusion Draft placeholders. I also observed the Timeline section is no longer included in some recent charters (notably GPU for the Web WG), so I dropped it.

>have the prospects of the Model API integration in the charter evolved given the recent prototyping efforts?

The Model Loader API is still considered an incubation and its tentative specification status would allow it to advance should the conditions enumerated in the note be satisfied. I will loop in @jbingham to confirm the latest status of the prototype.

wchao1115 commented on charter.html:
```diff
@@ -237,11 +255,11 @@ <h3 id="out-of-scope">
-            the Coordination section for details.
+            To avoid overlap with existing work, alignment with the Basic
+            Linear Algebra Subprograms (BLAS) interface is out of scope. The
+            WebGL and WebGPU shaders and WebAssembly SIMD are expected to
```
Since we already removed WebGL interop from the spec, should we also remove it here from the charter? Preferably noted as out of scope.

anssiko replied:
Thanks for the suggestions. I pushed https://github.com/w3c/machine-learning-charter/pull/19/commits/33e2eb59d47a8fcc6d722dbd9c1475c663a2900c that should address your comments. Namely:

I removed the WebGL reference from this BLAS compatibility context and added the following text following this paragraph:

>Interoperability between the WebNN and WebGL APIs is out of scope.

With this change, the only other reference to WebGL in addition to the above out of scope statement remains in the Motivation and Background section as follows:

>Currently, machine learning inference in the browser uses the WebGL graphics API with limited or no access to platform capabilities beneficial for ML such as CPU parallelism, general-purpose GPU, or dedicated ML hardware accelerators.

I think this is appropriate for context setting and to give a historical perspective. With WebNN API becoming more broadly available, the need to use WebGL API for ML workloads is expected to diminish over time.

wchao1115 commented on charter.html:
```diff
@@ -237,11 +255,11 @@ <h3 id="out-of-scope">
             algorithms.
```
We excluded training capability from the original charter. Do we want to rediscuss that aspect of the spec for webnn v2?

anssiko replied:
Thanks for raising this. I have opened #27 to solicit input on training capability. Please provide your feedback in that issue to feed into a broader WG discussion. We still have time to in-scope this if there is adequate interest.

anssiko committed:
Address review comments

- Remove reference to WebGL from the BLAS compatibility context
- Declare interop between the WebNN and WebGL APIs out of scope

anssiko committed:
Clarify Scope

Move examples of well-known model architectures from the bullet
list into the text section that talks about priority use cases.
Add transformers as another example.

Grammar fix: s/Allow/Allows

Fix #23

anssiko commented:
I met with @dontcallmedom and we merged the charter 2023-2025 PR. Dom is doing his final pass review before the charter starts its AC review.

anssiko commented:
(W3M & horizontal review in process: https://github.com/w3c/strategy/issues/367)
